### ISSUE TITLE
syslog-ng 3.19 crashes on reaching ~50MB

### ISSUE THREAD

**Issue by DivyaRamani15** [syslog-ng/syslog-ng#3350]
Hi Team,
We are facing an issue with syslog-ng-3.19.1 that came with the native SDK for our project.
We get to see that the syslog-ng reaches about ~58MB (not 58MB everytime, it is mostly between 42MB to 58MB),
and the process dies. It is then restarted by our application monitor script.

We were able to fetch the core dump for the same, (given below the complete decode using gdb).
The backtrace seems to point out "epoll_pwait" and the source being ivykis.

Please do help us with this, and do let us know if you think this is fixed in latest versions (FYI, we noticed
memory leak in 3.22 version). Also, please do let me know if you will need anything in addition to the coredump,
and the config file of ours.

Thanks in advance!
Divya.

Core generated by: /sbin/syslog-ng --cfgfile=/var/syslog-ng.conf
**The conf file:**
---------------
```
# THIS FILE WAS AUTO-GENERATED. DO NOT EDIT.

@version: 3.19
@include "scl.conf"

source s_local {
    #system();
    #internal();
    syslog(port(514) transport("udp"));
};

rewrite tag_SN {
    set("mAX-8BP3W9455100062", value("HOST"));
};

destination d_Log_Server {
    network("dev-syslog.surfboardconnect.com"
            port(1514)
            transport(tcp)

            time-zone("+00:00")
            template("${R_YEAR}-${R_MONTH}-${R_DAY}T${R_HOUR}:${R_MIN}:${R_SEC}.${R_MSEC}${TZOFFSET} ${HOST} ${MSGHDR}${MESSAGE}\n")
            );
};

log {
    source(s_local);
    rewrite(tag_SN);
    destination(d_Log_Server);
};

```
**GDB decode:**
-----------

```
GDB:
GNU gdb (GDB) 7.12.1
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law. Type "show copying"
and "show warranty" for details.
This GDB was configured as "--host=i686-pc-linux-gnu --target=arm-buildroot-linux-gnueabi".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
syslog-ng: No such file or directory.
[New LWP 20290]
[New LWP 20379]
Core was generated by `/sbin/syslog-ng --cfgfile=/var/syslog-ng.conf'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0 0xf6dbb79c in ?? ()
[Current thread is 1 (LWP 20290)]

(gdb) bt
#0 0xf6dbb79c in epoll_pwait () from /export/dramani1/wan-down-3198/build_WR11AX/targets/WR11AX/fs.install/lib/libc.so.6
#1 0xf6dbba18 in epoll_wait () from /export/dramani1/wan-down-3198/build_WR11AX/targets/WR11AX/fs.install/lib/libc.so.6
#2 0xf7092334 in iv_fd_epoll_timerfd_poll (st=0x80a7e0, active=0xffed49a8, abs=0x0) at iv_fd_epoll.c:436
#3 0xf708fe54 in iv_fd_poll_and_run (st=st@entry=0x80a7e0, abs=0xf70ca240 <stats_timer>) at iv_fd.c:202
#4 0xf7090a64 in iv_main () at iv_main_posix.c:112
#5 0xf7059570 in main_loop_run (self=0xf70c9d34 <main_loop>) at lib/mainloop.c:675
#6 0x000116fc in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
(gdb) info threads
Id Target Id Frame
* 1 LWP 20290 0xf6dbb79c in epoll_pwait ()
from /export/dramani1/wan-down-3198/build_WR11AX/targets/WR11AX/fs.install/lib/libc.so.6
2 LWP 20379 0xf6dbb79c in epoll_pwait ()
from /export/dramani1/wan-down-3198/build_WR11AX/targets/WR11AX/fs.install/lib/libc.so.6

(gdb) thread apply all bt

Thread 2 (LWP 20379):
#0 0xf6dbb79c in epoll_pwait () from /export/dramani1/wan-down-3198/build_WR11AX/targets/WR11AX/fs.install/lib/libc.so.6
#1 0xf6dbba18 in epoll_wait () from /export/dramani1/wan-down-3198/build_WR11AX/targets/WR11AX/fs.install/lib/libc.so.6
#2 0xf7092334 in iv_fd_epoll_timerfd_poll (st=0xf60005b8, active=0xf6952428, abs=0x818404) at iv_fd_epoll.c:436
#3 0xf708ff30 in iv_fd_poll_and_run (st=st@entry=0xf60005b8, abs=0x818404) at iv_fd.c:206
#4 0xf7090a64 in iv_main () at iv_main_posix.c:112
#5 0xf708ef48 in iv_work_thread (_thr=0x8183e0) at iv_work.c:185
#6 0xf70914b4 in iv_thread_handler (_thr=0x839340) at iv_thread_posix.c:112
#7 0xf6e31f14 in start_thread () from /export/dramani1/wan-down-3198/build_WR11AX/targets/WR11AX/fs.install/lib/libpthread.so.0
#8 0xf6dbb658 in ?? () from /export/dramani1/wan-down-3198/build_WR11AX/targets/WR11AX/fs.install/lib/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 1 (LWP 20290):
#0 0xf6dbb79c in epoll_pwait () from /export/dramani1/wan-down-3198/build_WR11AX/targets/WR11AX/fs.install/lib/libc.so.6
#1 0xf6dbba18 in epoll_wait () from /export/dramani1/wan-down-3198/build_WR11AX/targets/WR11AX/fs.install/lib/libc.so.6
#2 0xf7092334 in iv_fd_epoll_timerfd_poll (st=0x80a7e0, active=0xffed49a8, abs=0x0) at iv_fd_epoll.c:436
#3 0xf708fe54 in iv_fd_poll_and_run (st=st@entry=0x80a7e0, abs=0xf70ca240 <stats_timer>) at iv_fd.c:202
#4 0xf7090a64 in iv_main () at iv_main_posix.c:112
#5 0xf7059570 in main_loop_run (self=0xf70c9d34 <main_loop>) at lib/mainloop.c:675
#6 0x000116fc in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

(gdb) thread apply all bt full

Thread 2 (LWP 20379):
#0 0xf6dbb79c in epoll_pwait () from /export/dramani1/wan-down-3198/build_WR11AX/targets/WR11AX/fs.install/lib/libc.so.6
No symbol table info available.
#1 0xf6dbba18 in epoll_wait () from /export/dramani1/wan-down-3198/build_WR11AX/targets/WR11AX/fs.install/lib/libc.so.6
No symbol table info available.
#2 0xf7092334 in iv_fd_epoll_timerfd_poll (st=0xf60005b8, active=0xf6952428, abs=0x818404) at iv_fd_epoll.c:436
run_timers = 1
batch = 0xf69523d8
ret = <optimized out>
i = <optimized out>
#3 0xf708ff30 in iv_fd_poll_and_run (st=st@entry=0xf60005b8, abs=0x818404) at iv_fd.c:206
active = {next = 0xf6952428, prev = 0xf6952428}
run_timers = <optimized out>
#4 0xf7090a64 in iv_main () at iv_main_posix.c:112
_abs = {tv_sec = 0, tv_nsec = 0}
abs = <optimized out>
st = 0xf60005b8
run_timers = <optimized out>
#5 0xf708ef48 in iv_work_thread (_thr=0x8183e0) at iv_work.c:185
thr = 0x8183e0
pool = 0x809900
#6 0xf70914b4 in iv_thread_handler (_thr=0x839340) at iv_thread_posix.c:112
thr = 0x839340
#7 0xf6e31f14 in start_thread () from /export/dramani1/wan-down-3198/build_WR11AX/targets/WR11AX/fs.install/lib/libpthread.so.0
No symbol table info available.
#8 0xf6dbb658 in ?? () from /export/dramani1/wan-down-3198/build_WR11AX/targets/WR11AX/fs.install/lib/libc.so.6
No symbol table info available.
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 1 (LWP 20290):
#0 0xf6dbb79c in epoll_pwait () from /export/dramani1/wan-down-3198/build_WR11AX/targets/WR11AX/fs.install/lib/libc.so.6
No symbol table info available.
#1 0xf6dbba18 in epoll_wait () from /export/dramani1/wan-down-3198/build_WR11AX/targets/WR11AX/fs.install/lib/libc.so.6
No symbol table info available.
#2 0xf7092334 in iv_fd_epoll_timerfd_poll (st=0x80a7e0, active=0xffed49a8, abs=0x0) at iv_fd_epoll.c:436
run_timers = 0
--Type <return> to continue, or q <return> to quit--return
batch = 0xffed48d8
ret = <optimized out>
i = <optimized out>
#3 0xf708fe54 in iv_fd_poll_and_run (st=st@entry=0x80a7e0, abs=0xf70ca240 <stats_timer>) at iv_fd.c:202
active = {next = 0xffed49a8, prev = 0xffed49a8}
run_timers = <optimized out>
#4 0xf7090a64 in iv_main () at iv_main_posix.c:112
_abs = {tv_sec = 8443136, tv_nsec = 1}
abs = <optimized out>
st = 0x80a7e0
run_timers = <optimized out>
#5 0xf7059570 in main_loop_run (self=0xf70c9d34 <main_loop>) at lib/mainloop.c:675
No locals.
#6 0x000116fc in ?? ()
No symbol table info available.
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
(gdb)
```

**Comment by szemere** [comment]
Hello @DivyaRamani15 ,
 thank you for reporting this issue. Your Syslog-ng configuration looks normal to me. (I mean we have similar test cases in our CI system.) Can you share some details about the system your Syslog-ng is running on? You mentioned some SDK. I assume it might be an embedded system.

Best regards,
Laci

**Comment by DivyaRamani15** [comment]
Thanks for the response @szemere. The device we use is a mesh router, and Linux version 4.1.52.
And the SDK i mentioned is a broadcom CMS software developement code base (yes, an embedded system SDK) along with which we received this package of syslog.

**Comment by szemere** [comment]
Thank You for the quick response.
First I would like to confirm the origin of Syslog-ng. (Is it installed from an official repository? Built from source? Is there any internal patches above the official released version?)

note: If you do not have control over it:
- [Syslog-ng debun](https://www.syslog-ng.com/technical-documents/doc/syslog-ng-open-source-edition/3.16/administration-guide/87#syslog-ng-debun.1) will print as many information as possible, please investigate the possible arguments for privacy options. (i.e.: -l )
- If you start syslog-ng with the following command: `syslog-ng --version` it will print some information too.

**Comment by DivyaRamani15** [comment]
TMK, the syslog-ng comes as tar bundle (which should be from the official repository only), and we untar and build it with our toolchain as part of the firmware build, and there is no separate patches on top of it.
FYI:
# syslog-ng --version
syslog-ng 3 (3.19.1)
Config version: 3.19
Installer-Version: 3.19.1
Revision:
Compile-Date: Jun 23 2020 12:58:19
Module-Directory: /lib/syslog-ng
Module-Path: /lib/syslog-ng
Include-Path: /share/syslog-ng/include
Available-Modules: add-contextual-data,affile,afprog,afsocket,afuser,appmodel,basicfuncs,cef,confgen,cryptofuncs,csvparser,date,dbparser,disk-buffer,examples,graphite,hook-commands,json-plugin,kvformat,linux-kmsg-format,map-value-pairs,pseudofile,snmptrapd-parser,stardate,syslogformat,system-source,tags-parser,tfgetent,xml
Enable-Debug: off
Enable-GProf: off
Enable-Memtrace: off
Enable-IPv6: on
Enable-Spoof-Source: off
Enable-TCP-Wrapper: off
Enable-Linux-Caps: off
Enable-Systemd: off

**Comment by szemere** [comment]
Hello @DivyaRamani15,
 currently I am experimenting with Syslog-ng on an ARM platform. (It is a RaspberryPi. Not the best, but at least ARM.)
 I am running it with valgrind to spot any difference in memory allocation. Later I will try to run our tests with a memory limit.

Best regards,
Laci

**Comment by DivyaRamani15** [comment]
Hey @szemere , Thanks much!

**Comment by DivyaRamani15** [comment]
Hello @szemere , Any luck with the valgrind test?

**Comment by szemere** [comment]
Hello @DivyaRamani15 ,
 sorry for the delay. I encountered several problems during testing with valgrind. Some of them:

 - https://bugs.launchpad.net/raspbian/+bug/1857741
 - https://stackoverflow.com/questions/17430731/valgrind-returning-an-unhandled-instruction-on-raspberry-pi
 - https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=928224

Eventually I compiled valgrind for myself, and now most of the problems are eliminated.

Valgrind reported some uninitialized parameters around ivykis epoll calls. Since they are "behind" some unrecognized functions, I don't know yet if those findings are valid or not. Currently I am trying to figure out those issues.

Best regards,
Laci

**Comment by DivyaRamani15** [comment]
Thanks for the update @szemere. Not sure if those params and functions differ in latest version of ivykis. May be if some are already resolved in latest versions, but I'm not aware of those!

**Comment by szemere** [comment]
Hello,
 really good idea!
 Now, I compiled the latest released ivykis from the source code. No difference from the older one. :( Keep working.

Br,
Laci

**Comment by DivyaRamani15** [comment]
Hi @szemere , any clues on the leak / is it something that exists in the version for arm ?

Regards,
Divya.

**Comment by szemere** [comment]
Hello @DivyaRamani15,
 sorry for not answering earlier. I think I have tried everything I can to reproduce the issue, or try to find any difference between ARM and the x86 architectures which can cause it. Checked warnings and other Valgrind findings. (also running syslog-ng under memory limit)
 My next idea is: Since we do not really know your platform and we do not have access to it for experimenting we have two options:
A.) Giving you instructions for debugging ideas. This is a gray area and there is no guarantee for success.
B.) You might try to reproduce the issue in a "clear" qemu or other virtual environment. Once it is finished I can start it too, so I can experiment on it, trying to fix the issue.

Best regards,
Laci
